File: models/creator-card.js
const { ModelSchema, SchemaTypes, DatabaseModel } = require('@app-core/mongoose');

const modelName = 'creatorCards';

/**
 * @typedef {Object} CreatorCardSchema
 * @property {String} _id
 * @property {String} title
 * @property {String} [description]
 * @property {String} slug
 * @property {String} creator_reference
 * @property {Array<{title: String, url: String}>} [links]
 * @property {{currency: String, rates: Array<{name: String, description: String, amount: Number}>}} [service_rates]
 * @property {String} status
 * @property {String} access_type
 * @property {String} [access_code]
 * @property {Number} created
 * @property {Number} updated
 * @property {Number|null} deleted
 */

const schemaConfig = {
  _id: { type: SchemaTypes.ULID, required: true },
  title: { type: SchemaTypes.String, required: true },
  description: { type: SchemaTypes.String },
  slug: { type: SchemaTypes.String, unique: true, index: true },
  creator_reference: { type: SchemaTypes.String, required: true, index: true },
  links: { type: SchemaTypes.Mixed }, // Array of objects
  service_rates: { type: SchemaTypes.Mixed }, // Currency and rates array
  status: { type: SchemaTypes.String, required: true, index: true }, // 'draft' | 'published'
  access_type: { type: SchemaTypes.String, required: true, default: 'public' }, // 'public' | 'private'
  access_code: { type: SchemaTypes.String, default: null },
  created: { type: SchemaTypes.Number, required: true },
  updated: { type: SchemaTypes.Number, required: true },
  deleted: { type: SchemaTypes.Number, default: null },
};

const modelSchema = new ModelSchema(schemaConfig, { collection: modelName });

// Compound index to optimise the most frequent query: look up an active card by slug
// Both retrieval and delete queries filter on { slug, deleted } simultaneously
modelSchema.index({ deleted: 1, slug: 1 });

module.exports = DatabaseModel.model(modelName, modelSchema);

File: specs/creator-cards/data/creator-card.go
import ../../commons.go

CreatorCard {
  _id string<isUnique|indexed> // Unique identifier (ULID)
  title string // Title of the creator card
  description? string // Optional description of the creator card
  slug string<isUnique|indexed> // Unique slug for URL routing
  creator_reference string<indexed> // Creator reference string
  links[]? { // Optional list of social links
    title string // Title of the link
    url string // URL of the link
  }
  service_rates? { // Optional service rates configuration
    currency string // Currency code (e.g. NGN, USD, GBP, KES, GHS)
    rates[] { // Array of rates for services
      name string // Name of the rate service
      description string // Description of the rate service
      amount number // Amount in smallest unit of currency (integer)
    }
  }
  status string(draft|published) // Card status: draft, published
  access_type string(public|private) // Access type: public, private. Defaults to public
  access_code? string // Alphanumeric access code of exactly 6 characters, required on private cards
  ...common
}
